### PR TITLE
Upgrade lwc-jest, add unit tests

### PR DESCRIPTION
**Pull request by trevor-bliss**

### CODE REVIEW

File: force-app/test/jest-mocks/lightning/navigation.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+/**
+ * For the original lightning/navigation mock that comes by default with
```
Comment (muenzpraeger) [inline force-app/test/jest-mocks/lightning/navigation.js]: Will the default mock that's shipped with `lwc-jest` make this obsolete in the future? Or is this an enhancement to the shipped mock that'll go eventually into `lwc-jest`?

Comment (trevor-bliss) [reply force-app/test/jest-mocks/lightning/navigation.js]: This is an enhancement to what's shipped with `lwc-jest`. I don't plan on adding these changes to the default mock unless there's strong support by others and customers to do so.

File: force-app/main/default/lwc/placeholder/__tests__/placeholder.test.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        }
+    });
+
+    it('sets img url to be bike_assets resource', () => {
+        const element = createElement('c-placeholder', {
+            is: Placeholder,
+        });
+        document.body.appendChild(element);
+        const img = element.shadowRoot.querySelector('img');
+        // By default @salesforce/lwc-jest will resolve the
```
Comment (kevinv11n) [inline force-app/main/default/lwc/placeholder/__tests__/placeholder.test.js]:
```suggestion
        // By default @salesforce/lwc-jest resolves the
```

File: force-app/main/default/lwc/placeholder/__tests__/placeholder.test.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        // @salesforce/resourceUrl/bike_assets import to be "bike_assets"
```
Comment (kevinv11n) [inline force-app/main/default/lwc/placeholder/__tests__/placeholder.test.js]:
```suggestion
        // @salesforce/resourceUrl/bike_assets import to "bike_assets"
```

File: force-app/main/default/lwc/productFilter/__tests__/productFilter.test.js [hunk 4]
```diff
@@ -132,4 +138,42 @@ describe('c-product-filter', () => {
...
     });
+
+    describe('getPicklistValues @wire error', () => {
+        it('shows error message elements', () => {
+            const element = createElement('c-product-filter', {
+                is: ProductFilter,
+            });
+            document.body.appendChild(element);
+
+            getPicklistValuesAdapter.error();
```
Comment (kevinv11n) [inline force-app/main/default/lwc/productFilter/__tests__/productFilter.test.js]: Nice!

File: force-app/main/default/lwc/productListItem/__tests__/productListItem.test.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    });
+
+    it('navigates to record page when View Details button clicked', () => {
+        const expectedId = 'expectedId';
+        const element = createElement('c-product-list-item', {
+            is: ProductListItem,
+        });
+        element.product = {
+            Id: expectedId,
+            Picture_URL__c: 'https://example.com',
```
Comment (kevinv11n) [inline force-app/main/default/lwc/productListItem/__tests__/productListItem.test.js]:
```suggestion
            Picture_URL__c: 'https://salesforce.com',
```

File: force-app/main/default/lwc/productTile/__tests__/productTile.test.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            is: ProductTile,
+        });
+        // emulate a DragEvent, jsdom does not implement this class yet
+        const dragstartEvent = new CustomEvent('dragstart');
+        dragstartEvent.dataTransfer = {
+            setData: jest.fn(),
+        };
+        const product = {
+            Id: 1,
+            Picture_URL__c: 'https://example.com',
```
Comment (kevinv11n) [inline force-app/main/default/lwc/productTile/__tests__/productTile.test.js]:
```suggestion
            Picture_URL__c: 'https://salesforce.com',
```

File: force-app/main/default/lwc/productTile/__tests__/productTile.test.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+    it('clicking fires selected event', () => {
+        const listener = jest.fn();
+        const element = createElement('c-product-tile', {
+            is: ProductTile,
+        });
+        element.addEventListener('selected', listener);
+        element.product = {
+            Id: 1,
+            Picture_URL__c: 'https://example.com',
```
Comment (kevinv11n) [inline force-app/main/default/lwc/productTile/__tests__/productTile.test.js]:
```suggestion
            Picture_URL__c: 'https://salesforce.com',
```

File: force-app/main/default/lwc/productTileList/__tests__/productTileList.test.js [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+// Mock out the pubsub lib and use these mocks to verify how functions were called
+jest.mock('c/pubsub', () => {
+    return {
+        fireEvent: jest.fn(),
+        registerListener: jest.fn(),
+        unregisterAllListeners: jest.fn(),
+    };
+});
+
+// realistic data with multiple records
```
Comment (kevinv11n) [inline force-app/main/default/lwc/productTileList/__tests__/productTileList.test.js]:
```suggestion
// Realistic data with multiple records
```

File: force-app/main/default/lwc/productTileList/__tests__/productTileList.test.js [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+const mockGetProducts = require('./data/getProducts.json');
+// an empty list of records to verify the component does something reasonable
```
Comment (kevinv11n) [inline force-app/main/default/lwc/productTileList/__tests__/productTileList.test.js]:
```suggestion
// An empty list of records to verify the component does something reasonable
```

File: force-app/main/default/lwc/productFilter/__tests__/productFilter.test.js [hunk 10]
```diff
@@ -132,4 +138,42 @@ describe('c-product-filter', () => {
...
+
+            return Promise.resolve().then(() => {
+                const messages = element.shadowRoot.querySelectorAll(
+                    'c-inline-message',
+                );
+                // one error message per @wire
```
Comment (kevinv11n) [inline force-app/main/default/lwc/productFilter/__tests__/productFilter.test.js]:
```suggestion
                // One error message per @wire
```

File: force-app/main/default/lwc/productListItem/__tests__/productListItem.test.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        };
+        document.body.appendChild(element);
+
+        const lightningButton = element.shadowRoot.querySelector(
+            'lightning-button',
+        );
+        lightningButton.click();
+
+        const { pageReference } = getNavigateCalledWith();
+        // verify the component under test called the correct navigate event
```
Comment (kevinv11n) [inline force-app/main/default/lwc/productListItem/__tests__/productListItem.test.js]:
```suggestion
        // Verify the component under test called the correct navigate event
```

File: force-app/main/default/lwc/productTile/__tests__/productTile.test.js [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        while (document.body.firstChild) {
+            document.body.removeChild(document.body.firstChild);
+        }
+    });
+
+    it('dragging sets product as dataTransfer data', () => {
+        const element = createElement('c-product-tile', {
+            is: ProductTile,
+        });
+        // emulate a DragEvent, jsdom does not implement this class yet
```
Comment (kevinv11n) [inline force-app/main/default/lwc/productTile/__tests__/productTile.test.js]:
```suggestion
        // Emulate a DragEvent, jsdom does not implement this class yet
```

File: force-app/test/jest-mocks/lightning/navigation.js [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+            _replace = replace;
+        }
+        [GenerateUrl](pageReference) {}
+    };
+};
+NavigationMixin.Navigate = Navigate;
+NavigationMixin.GenerateUrl = GenerateUrl;
+
+/*
+ * Tests will not have access to the internals of this mixin used by the
```
Comment (kevinv11n) [inline force-app/test/jest-mocks/lightning/navigation.js]:
```suggestion
 * Tests do not have access to the internals of this mixin used by the
```

File: force-app/test/jest-mocks/lightning/navigation.js [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+ * component under test so we save a reference to what the Navigate method is
```
Comment (kevinv11n) [inline force-app/test/jest-mocks/lightning/navigation.js]:
```suggestion
 * component under test so save a reference to the arguments the Navigate method is
```

File: force-app/test/jest-mocks/lightning/navigation.js [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+ * called with and provide access to the params in the test via this function.
```
Comment (kevinv11n) [inline force-app/test/jest-mocks/lightning/navigation.js]:
```suggestion
 * invoked with and provide access with this function.
```

**Pull request merged**